src/zenml/integrations/kubernetes/deployers/kubernetes_deployer.py

```diff
@@ -0,0 +1,1009 @@
+                f"Cleanup failed: {cleanup_error}. "
+                f"Manual cleanup: kubectl delete all,configmap,secret "
+                f"-n {ctx.namespace} -l {label_selector}"
+            )
+
+    def _clear_deployment_context(self) -> None:
+        """Clear deployment context to prevent state pollution between operations."""
+        self._ctx = None
+        self._engine = None
+
+    def _initialize_deployment_context(
+        self, deployment: DeploymentResponse
+    ) -> None:
+        """Initialize basic deployment context from deployment snapshot.
+
+        Sets up instance variables: _settings, _namespace, _resource_name,
+        _labels, and _engine (if needed).
+
+        Args:
+            deployment: The deployment to initialize context for.
+
+        Raises:
+            DeployerError: If deployment has no snapshot.
+        """
+        self._clear_deployment_context()
+
+        snapshot = deployment.snapshot
+        if not snapshot:
+            raise DeployerError(
+                f"Deployment '{deployment.name}' has no snapshot"
+            )
+
+        settings = cast(
+            KubernetesDeployerSettings, self.get_settings(snapshot)
+        )
+        namespace = settings.namespace or self.config.kubernetes_namespace
+        resource_name = kube_utils.sanitize_label(f"zenml-{deployment.id}")[
+            :MAX_K8S_NAME_LENGTH
+        ]
+        secret_name = resource_name
+
+        labels = {
+            "zenml-deployment-id": str(deployment.id),
+            "zenml-deployment-name": kube_utils.sanitize_label(
+                deployment.name
+            ),
+            "managed-by": "zenml",
+        }
+        if settings.labels:
+            labels.update(**settings.labels)
+
+        image = self.get_image(snapshot)
+
+        self._ctx = _DeploymentCtx(
+            settings=settings,
+            namespace=namespace,
+            resource_name=resource_name,
+            secret_name=secret_name,
+            labels=labels,
+            image=image,
+        )
+
+    def _prepare_deployment_resources(
+        self,
+        deployment: DeploymentResponse,
+        environment: Dict[str, str],
+        secrets: Dict[str, str],
+    ) -> List[Dict[str, Any]]:
+        """Prepare all resources needed for deployment.
+
+        Args:
+            deployment: The deployment to prepare.
+            environment: Environment variables.
+            secrets: Secret environment variables.
+
+        Returns:
+            List of rendered kubernetes resources.
+
+        Raises:
+            DeploymentProvisionError: If preparation fails.
+        """
+        snapshot = deployment.snapshot
+        assert snapshot, "Pipeline snapshot not found"
+
+        self._initialize_deployment_context(deployment)
+        ctx = self.require_ctx()
+
+        self._engine = KubernetesTemplateEngine(
+            custom_templates_dir=ctx.settings.custom_templates_dir
+            or self.config.custom_templates_dir
```

Comment:
You don't need to check for both. The ZenML core will merge config into settings for you.


```suggestion
            custom_templates_dir=ctx.settings.custom_templates_dir
```